### PR TITLE
Feature: Porting unique index for AO table to CBDB

### PR DESCRIPTION
**Pull request by lss602726449**
closes: #68 

---

### Change logs
In this PR, we mainly Port unique index for AO/AOCS table to CBDB.
Since the Memtuple in the AO table does not have the xmin and xmax attributes, we realize unique check by using its auxiliary index structure BlockDirectory (Heaptuple). Specifically, for the scenario with a unique index, when inserting into the AO table, we insert a placeholder in BlockDirectory to block the insertion of the same key. Its key design logic, see [unique-index README](https://github.com/lss602726449/cloudberrydb/blob/unique_index/src/backend/access/appendonly/README.md#unique-indexes).
Specific changes include:
1. Make changes to BlockDirectory. The previous BlockDirectory structure cannot support unique index, and the fill hole mechanism is removed.
2. Realize the insert , delete and update of the AO table in the unique index scenario
3. Add a new field version to the pg_appendonly table to distinguish it from the previous BlockDirectory structure
4. Change the vacuum implementation logic of the AO table and index. When the number of dead tuples in the segment reaches the threshold, drop the segment and recycle the segment and the corresponding index.

### Why are the changes needed?
1. the previous BlockDirectory can not support unique index check
image
2. Change the vacuum implementation for more efficient vacuum

### Does this PR introduce any user-facing change?
introduce for unique index for AO/AOCS
```
create table test (a int unique) using ao_row;
insert into test select generate_series(1,5);
insert into test values(1); --Error: duplicate key
delete from test where a = 1;
insert into test values(1); --success
update set a = 1 where a = 1; --success
vacuum test; --vacuum the segment and index 
```

### How was this patch tested?

The main test cases are almost all cherry-picked from GP , mainly including:
isolation2: ao_blkdir, ao_unique_index, aocs_unique_index, ao_unique_index_vacuum
regress: uao_dml_unique_index_update, uao_dml_unique_index_delete, ao_unique_index_build

our changes include: ao_unique_index_vacuum
Change the test of verifying index vacuum to sequential execution. There are inconsistencies under distributed transactions during concurrent execution. For details, see bugs behind and file vaccum_ao.c
```
-- psql 1
CREATE EXTENSION gp_inject_fault;
CREATE TABLE unique_index_vacuum_ao_row(i int UNIQUE) USING ao_row
    DISTRIBUTED REPLICATED;
INSERT INTO unique_index_vacuum_ao_row SELECT generate_series(1, 5);
DELETE FROM unique_index_vacuum_ao_row WHERE i = 5;
SELECT gp_inject_fault('vacuum_ao_after_index_delete', 'suspend', dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content <> -1;

vacuum unique_index_vacuum_ao_row;     --execute behind the gp_wait_until_triggered_fault
```
```
--psql2
SELECT gp_wait_until_triggered_fault('vacuum_ao_after_index_delete', 1, dbid)
    FROM gp_segment_configuration WHERE role = 'p' AND content <> -1;
```
image

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [x] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [x] Sign the Contributor License Agreement as prompted for your first-time contribution.
- [x] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [x] Document changes.
- [x] Add tests for the change
- [x] Pass `make installcheck`
- [x] Pass `make -C src/test installcheck-cbdb-parallel`
- [x] Feel free to @cloudberrydb/dev team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/backend/access/appendonly/appendonly_blkdir_udf.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
+/*------------------------------------------------------------------------------
+ *
+ * AppendOnly_Blkdir UDFs
+ *   User-defined functions (UDF) for support of append-only block directory
+ *
+ * Copyright (c) 2013-Present VMware, Inc. or its affiliates.
```
Comment (yjhjstz) [inline src/backend/access/appendonly/appendonly_blkdir_udf.c:6]: use cbdb's Copyright

File: src/include/catalog/pg_appendonly.h [hunk 2]
```diff
@@ -34,6 +34,7 @@ CATALOG(pg_appendonly,6105,AppendOnlyRelationId)
     bool            columnstore;        /* true if orientation is column */ 
     Oid             segrelid;           /* OID of aoseg table; 0 if none */
     int16           segfilecount;		/* the (per seg) average total number of segment file */
+	int16			version;			/* AO relation version */
```
Comment (yjhjstz) [inline src/include/catalog/pg_appendonly.h:37]: align code

File: src/backend/access/appendonly/appendonly_blkdir_udf.c (hunk 1 above)
Comment (tuhaihe) [reply src/backend/access/appendonly/appendonly_blkdir_udf.c:6]: I found the Cloudberry Database team does not initially create or change this file, so we should keep the original copyright information unchanged.

File: src/include/catalog/pg_appendonly.h (hunk 2 above)
Comment (lss602726449) [reply src/include/catalog/pg_appendonly.h:37]: Thanks for pointing out，I'll fix in the next commit

**Comment by yjhjstz** [comment]
support `ADD CONSTRAINT` ?
```
ALTER TABLE table_name ADD CONSTRAINT constraint_name UNIQUE (column_name);
```

**Comment by lss602726449** [comment]
> 



> support `ADD CONSTRAINT` ?
> 
> ```
> ALTER TABLE table_name ADD CONSTRAINT constraint_name UNIQUE (column_name);
> ```

Sure, almost use unique index same as heap, execpt different implementation.
image

**Comment by yjhjstz** [comment]
> > support `ADD CONSTRAINT` ?
> > ```
> > ALTER TABLE table_name ADD CONSTRAINT constraint_name UNIQUE (column_name);
> > ```
> 
> Sure, almost use unique index same as heap, execpt different implementation. image

add some test like `create unique index pkeys_i on pkeys (pkey1, pkey2);`

**Comment by lss602726449** [comment]
> > Sure, almost use unique index same as heap, execpt different implementation. image
> 
> add some test like `create unique index pkeys_i on pkeys (pkey1, pkey2);`
Thanks for your suggestion. I will add more unique test about multiple key in the next commit
image

**Review by HuSen8891 — APPROVED** [review]

**Review by yjhjstz — DISMISSED** [review]

**Review by yjhjstz — APPROVED** [review]

**Review by HuSen8891 — APPROVED** [review]

**Pull request merged**